embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/anr/detection/ResponsivenessMonitor.kt

```diff
@@ -0,0 +1,106 @@
+package io.embrace.android.embracesdk.anr.detection
+
+import com.google.gson.annotations.SerializedName
+import io.embrace.android.embracesdk.internal.clock.Clock
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+ * Enables the monitoring of a component or thread's responsiveness by bucketing the time lapsed between calls of [ping] into the specified
+ * durational buckets separated by the given boundaries. The counts of the buckets are provided along with start and end times of the
+ * first N outliers, defined by any interval greater than [outlierThreshold], where N specified by [outliersLimit].
+ *
+ * The bucket name is the boundary value (inclusive) down to the next boundary value or 0. The last bucket is anything greater than the
+ * last boundary and less than [Long.MAX_VALUE].
+ *
+ * Note that this does not handle backpressure so don't use it if the rate of pings is expected to exceed the rate at which the pings can
+ * be processed sequentially.
+ */
+internal class ResponsivenessMonitor(
+    private val clock: Clock,
+    private val name: String,
+    private val outliersLimit: Int = 100,
+    private val outlierThreshold: Long = 500L,
+    boundaries: List<Long> = listOf(120L, 250L, 500L, 2000L)
+) {
+    private val buckets = boundaries.sorted().plus(Long.MAX_VALUE)
+    private val outliers = mutableListOf<Outlier>()
+    private val gaps: MutableMap<Long, Long> = buckets.associateWith { 0L } as MutableMap<Long, Long>
+    private val firstPing = AtomicLong(unsetPingTime)
+    private val latestPing = AtomicLong(unsetPingTime)
+    private val errors = AtomicLong(0)
+
+    @Synchronized
+    fun ping() {
+        val previousPing = latestPing.get()
+        latestPing.set(clock.now())
+        if (previousPing != unsetPingTime) {
+            (latestPing.get() - previousPing).let { gap ->
+                if (gap >= 0) {
```

Comment:
might be good to break this into a separate function - I found it hard to hold all the information in my brain at once

Reply:
Yup. I can do that. I tried using actual names instead of `it` to make it more readible, but this is still pretty terse. I'm sure I'll thank myself in 3 months when I look at this again.